### PR TITLE
Proposal - Switch to godep for dependency management

### PR DESCRIPTION
**Pull request by icecrime**
_This is a PoC to collect some feedback._
## Rationale

As Docker grows, more specialized teams take ownership of more specialized repositories: we've seen it with [docker/distribution](https://github.com/docker/distribution), more recently with [docker/libnetwork](https://github.com/docker/libnetwork), and everything points to a continuing trend (builder might be pulled out, etc.).

All this to say that vendoring is becoming a more and more relevant piece, and the current mechanism has one significant flaw: we vendor dependent repositories as a whole, rather than selectively copying the pieces we actually use. It is [manually done](https://github.com/docker/docker/blob/master/hack/vendor.sh#L63-L70) in some cases, but that is not going to scale.
## Solution

[Godep](https://github.com/tools/godep) has the nice property that it **only** vendors what's actually imported. However, it doesn't allow for fine-grained specification of the git hash or tag that we want to pull (something our `hack/vendor.sh` script allows).

What I attempted here is to take the best of both world.
- Continue specifying the dependency version in the `hack/vendor.sh` script.
- Use these references to pull down the entire dependent repositories into `vendor/`.
- Rely on `godep` to extract from those the pieces that we actually import.

To put it differently, we use the `vendor/` directory as a temporary "buffer" which holds the entire dependencies, before having `godep` move to `Godeps/_workspace` the parts we actually use.
## Result

Today this results in a net save of 32k lines.

```
 $ git diff origin/master --shortstat
 1685 files changed, 129891 insertions(+), 162372 deletions(-)
```

I'm sure that difference will grow (for example as we start relying on docker/distribution for the client-side code with #13375).
## Issues

I had to introduce 2 `go get` in the `Dockerfile`: one for `godep` (which I think is fine), another more problematic for `code.google.com/p/go.net/html`. This used to be handled by a `COPY vendor`, I'm not sure what's the best way to handle this now.

You probably want to review https://github.com/docker/docker/commit/81d0fee9ac3058664deb28f03d3df95ae9901be0 as the two others are removing `vendor/` and adding `Godeps/`. Let me know what you think, and ping the experts (@tianon @jfrazelle).

### CODE REVIEW

**Comment by icecrime** [comment]
Holy sh*t it's green.

**Comment by runcom** [comment]
Can I vote for gb (http://getgb.io/) instead of godep? :D (maybe you've already talked about this and I missed it)

**Comment by phemmer** [comment]
> is better "integrated" with standard go flows. 

I think gb is less "integrated" with standard go flows.

http://getgb.io/docs/project/

> gb projects do not live inside your $GOPATH

> gb does not use go get to manage a project’s dependencies, the owner of the project should copy any code that the project depends on

> gb projects can be retrieved using go get, but cannot be built by the go tool

So basically this would require manually adding and removing dependencies, and building the code would now require the `gb` utility.

**Comment by icecrime** [comment]
To answer some of the questions:

- I went with `godep` specifically for its ability to copy imported paths only rather than full repos (it [internally](https://github.com/tools/godep/blob/master/pkg.go#L38) relies on `go list -json` that we may as well use directly). At far as I know, `gb` doesn't do that.
- I tried getting deps from `docker/docker/docker` but it was very partial for some reason. We also have test specific dependencies.

**Comment by icecrime** [comment]
Super rewarding to work after hours on improving things to see them ditched without even a proper review...

**Comment by tianon** [comment]
I'm really, really sorry @icecrime! :cry:

Offense was not my intention in the least -- my reservations with this PR are the addition of new tools for managing dependencies.  I had a crazy idea that we might be able to accomplish this with `go list` alone, but I didn't want to have you waste a bunch of your time if it turned out to be a dead end, so I tried it out before responding here. :disappointed:

**Comment by tianon** [comment]
(I also did not intend to discount your work here in any way -- what you're attempting here is a noble goal IMO!)

**Comment by tianon** [comment]
I'm definitely curious why the removed lines here aren't higher -- any
ideas?  You mention in your first comment the 32k figure, but github should
be showing that same number.

**Comment by icecrime** [comment]
The crazy idea that I mentioned here https://github.com/docker/docker/pull/13756#issuecomment-109332170? Right, who would have thought.

**Comment by tiborvass** [comment]
Disclaimer: I hate Godep for I can never get my head around what commands to use.

But here are my more rational thoughts: if we can do things without a new dependency and is easy enough to maintain, we should choose the solution without the new dependency.

So to be clear, I agree with https://github.com/docker/docker/pull/13756#issuecomment-109332170 
> ...relies on `go list -json` that we may as well use directly

I understand both @icecrime and @tianon, but if you agree with what I said above, I would suggest reopening #13783.

Hopefully we can get this rolling. Thanks for looking into this

Now @docker/core-maintainers please comment and vote +1/-1.

**Comment by aluzzardi** [comment]
+1 for godep

`docker/docker` is the **only** docker project not using godep (okay, granted, it's by far the biggest project, but sitll).

**Comment by aluzzardi** [comment]
Also, regarding not allowing to specify a git hash - I've never found that to be a real problem for Swarm.

Dependencies do get checked in alongside the code, so the build is always reproducible no matter what.

Let's say you want to revert to a specific revision of `libcontainer`, for instance `v2.0.1`:
```
cd ~/go/src/github.com/docker/libcontainer
git checkout v2.0.1
cd ~/go/src/github.com/docker/docker
godep update github.com/docker/libcontainer
```

`Godeps/Godeps.json` will contain the **2.0.1** hash for future reference.
`Godeps/_workspace/src/github.com/docker/libcontainer` will contain the code for **2.0.1**

So I don't see what the problem of using godep straight up would be?

**Comment by jfrazelle** [comment]
I am not a fan of godep

**Comment by endophage** [comment]
+1 for Godep. Consistency and uniformity are worth a lot.

File: Dockerfile [hunk 1]
```diff
@@ -178,8 +178,13 @@ RUN set -x \
 	&& git clone https://github.com/BurntSushi/toml.git /go/src/github.com/BurntSushi/toml \
 	&& (cd /go/src/github.com/BurntSushi/toml && git checkout -q $TOMLV_COMMIT)
 
+# Download godep & jq
+RUN go get github.com/tools/godep
+RUN curl -sSL http://stedolan.github.io/jq/download/linux64/jq -o /usr/local/bin/jq \
```
Comment (stevvooe) [inline Dockerfile]: Use https here.

File: Dockerfile [hunk 2]
```diff
@@ -178,8 +178,13 @@ RUN set -x \
...
+   && chmod +x /usr/local/bin/jq
+
 # copy vendor/ because go-md2man needs golang.org/x/net
-COPY vendor /go/src/github.com/docker/docker/vendor
+RUN go get code.google.com/p/go.net/html
```
Comment (stevvooe) [inline Dockerfile]: This is a very old version of the dependency. Should be something like `golang.org/x/net/html`.

**Comment by tiborvass** [comment]
@stevvooe please don't review code, this is in design review.

**Comment by dmcgowan** [comment]
+1 Godeps

Not a fan of godeps and I find Tianon's change cleaner, but using godep is more consistent and diff suppression is very useful.

**Comment by jfrazelle** [comment]
I really don't like godeps

**Comment by stevvooe** [comment]
:+1: 

Right now, `godep` is the best existing dependency tool. It gets out of the way and ensures compatibility with `GOPATH`. Yes, it can be awful to work with and frustrating when adding or updating dependencies. Really, it works where needed -- during build. Day to day, I don't even have to use `godep`. I manage my GOPATH with `go get` and checkouts. When I am ready for testing, I run `GOPATH=$(godep path):$GOPATH make`.

`gb` looks like an interesting project but it is way too [opinionated](http://getgb.io/docs/project/) and departs from the Go package layout that makes building and the use of `go get` work well. Effectively, the way `gb` is laid out means that once you adopt that pattern, other projects that don't use `gb` cannot share that code.

**Comment by mrjana** [comment]
FWIW(because this issue is closed), +1 godep. I am not sure why folks hate godep. Care to explain the reasons since I am curious.

**Comment by phemmer** [comment]
@jfrazelle 

> I am not a fan of godep

&nbsp;
> I really don't like godeps

Any particular reasons?

I don't think godep is perfect. But I do think the positives outweigh the negatives.

**Comment by jfrazelle** [comment]
I think the solution with pure bash is cleaner and if we can do it without adding a new dependency or tool than we should

As for godeps I've tried to like it, I want to like it, but I really don't, it's just not user friendly and doesn't do the things I even want like designating my own sha 

It's far too magical, I like to control the sha/tag being used

**Comment by jfrazelle** [comment]
Plus bash is my favorite programming language so I'm going to play the insanely biased towards all bash card here ;)

**Comment by jfrazelle** [comment]
Also I had this absolutely horrific experience trying to get out out a godeps hell with our fork of drone when that was our ci, so it's definitely not perfect because I couldn't even get it to compile which is kinda the point no?

**Comment by tiborvass** [comment]
I echo @jfrazelle's experience, multiple times. I just retried it now, from scratch, and got `directory "/Users/tibor/go/src" is not using a known version control system` on a `godep save` following their README.

Also for everyone +1-ing Godep @aluzzardi @endophage @dmcgowan @stevvooe @mrjana @phemmer, just to be clear: the solution proposed in this PR is to continue using vendor.sh, but to rely on `godep` for the smart part of "smart vendoring", which like it was speculated, can be done easily with `go list`. So it wouldn't be like using `godep`, it would be like using `vendor.sh` which would use `godep` under the hood (unless I'm wrong then please correct me).

A big advantage of vendor.sh is that it specifies the exact hashes/tags needed for packagers, (that's actually why this PR mixes both). The argument I'm trying to make is that you don't need `godep` just for its smart vendoring capabilities since the smartness comes from `go list` not `godep`, and that's exactly what is used in #13783, in 30 lines of readable bash instead of a new dependency.

Last thing: currently, like it or not, our vendoring tool *is* `vendor.sh` and I'd prefer a small and simple improvement that results in removing a huge numbers of lines, versus having a tool war now.

**Comment by icecrime** [comment]
First of all, let me correct my initial post. I was wrong to think that Godep doesn't let us finely pick the git hashes we want: it appears [`godep restore` does just that](https://github.com/tools/godep/blob/master/restore.go#L19). That means the hybrid solution initially proposed here is probably not necessary, and that we end up with the simple choice of a homemade `go list` based solution, or going entirely with `godep`.

On this regard, I fail to understand what's wrong about adding a dependency on a tool that, like it or not, is widely known and used by our entire programming language community: this is like saying that you're gonna write Python but not use pip. So on my end, after clarifying that Godep allows to pick our git hashes, I don't see any argument left in favor of a custom script but NIH syndrome.

I'll leave it to the majority of maintainers to decide. Whatever the outcome is, I'd appreciate being asked to amend my PR as in any other design review rather than submitting concurrent implementation, because ultimately we *do* need more selective vendoring whatever tool we pick,

**Comment by aluzzardi** [comment]
@jfrazelle @tiborvass I personally never had an issue with godep. The very few times I had errors was either because I had a weird Go setup or because I did something wrong.

+1 @icecrime for godep only

Also, before everyone votes, I invite you to read the dependencies & vendoring proposal on goland-dev:
https://groups.google.com/forum/#!msg/golang-dev/nMWoEAG55v8/iJGgur7W_SEJ

To sum up:
- vendoring "standards" for Go are coming
- Actually, the standard library of Go 1.4 already uses them
- Go 1.5 will make them available to everyone
- Future versions of Go will also include a standard way to define the hashes you want
- To quote the proposal, *we’ll probably want something like godep’s JSON format*
- Go will not provide the tools to handle dependencies, only a standard
- The author of godep has already agreed in that thread to follow the new standard
- The proposal cites godep at least 5 times

So yeah, we can make a quick change now to use `go list` in `vendor.sh`, then when Go 1.5 comes out another quick change to use the `internal/` format, then when 1.6 comes out another "quick" change to parse JSON in bash so that `vendor.sh` can read godep files - or we can just switch to godep already.

My vote stays: Switch to godep or leave `vendor.sh` alone - all efforts to improve our own custom vendoring is a waste of time IMO.

**Comment by aluzzardi** [comment]
Just to clarify: I'm not saying `vendor.sh` is bad - it served us well for a long time, way before godep even existed.

However, I think we should follow best practices as they emerge.

**Comment by jfrazelle** [comment]
If the go team loves Godep so much then why would Dave Cheney make his own
tool

**Comment by jfrazelle** [comment]
If you can show me 2 projects by a single member of the go team where they
personally use Godep then I am sold

**Comment by tiborvass** [comment]
@icecrime Ok, thanks for clarifying that!

@aluzzardi I found the error with godep: your current repository has to be a (git?) repo for some reason. Once I did `git init` it worked, which is completely stupid because why couldn't I use a non-versioned folder as a Go project? It is not written in their README. Anyway at least now I can test it.

Which brings me to my second issue: you need to remove _workspace every time you use `godep restore`. Not sure if that's a bug or what but look: https://gist.github.com/tiborvass/68285abf1435d86273ae

The good thing though is that we can indeed use both tags and hashes in Godep.

And about the golang-dev thread: it is absolutely not clear what is coming for 1.5. Many people are divided on the issue including among the core team.

**Comment by mrjana** [comment]
@tiborvass Can `hack/vendor.sh` work with a non-versioned folder? IMO I think all sources of vendoring should be versioned and I don't think it is stupid to have that constraint. It's just a sensible way to avoid confusion.

**Comment by jfrazelle** [comment]
I agree that thread was not a good "pro godep" resource haha

**Comment by davecheney** [comment]
How can I help make gb work for docker?

**Comment by tianon** [comment]
@davecheney from what I've looked at so far, the two issues we'd run into are the `src/` mandate (which is going to be unpopular) and then the problem of our esoteric build flags for making static compilation work

I'd also say that `gb-vendor` isn't going to work for us since it doesn't support using upstream release tags instead of hashes (even if you hand-munge the JSON), but since it's an optional piece I don't see that as a blocker myself.

**Comment by icecrime** [comment]
@davecheney Ohai! My two concerns with `gb` as I understand it so far:

- It doesn't solve the initial concern of vendoring *specifically* what's imported. I understand it's likely not a must have for most projects, but it can make a different for a project the size of Docker (especially as we are taking a path where vendoring is going to take more and more importance).
- I don't think we can reasonably move docker sources around (as it would automatically require every outstanding PR to rebase). I'm not sure `gb` will allow to simultaneously override the `ProjectRoot` (`-R`) with the root of the repo *and* having vendored dependencies in a child directory of the said `ProjectRoot`.

Thanks for chiming in though ;-)

**Comment by davecheney** [comment]
> It doesn't solve the initial concern of vendoring specifically what's imported. I understand it's likely not a must have for most projects, but it can make a different for a project the size of Docker 

`gb vendor fetch URL -branch / -revision` 

> I don't think we can reasonably move docker sources around (as it would automatically require every outstanding PR to rebase). I'm not sure gb will allow to simultaneously override the ProjectRoot 

Yup, this is a cost. I'm afraid I cannot alleviate that. 

btw. @jfrazelle I don't work for Google, my views and my mistakes are entirely my own.

**Comment by davecheney** [comment]
> from what I've looked at so far, the two issues we'd run into are the src/ mandate (which is going to be unpopular) and then the problem of our esoteric build flags for making static compilation work

The gb project layout is what it is. Partly this is to give a strong separation between your code and your dependencies, which is informed by the strong dislike of rewriting source paths to munge it all into one blob. The other part is gb is deliberately different, it doesn't try to co-opt the Go tool into doing a slightly better job; that's been tried and it failed, by my count over two dozen times now. Making another slight variation on that theme would also be a market failure, so gb is more avant-garde.

> then the problem of our esoteric build flags for making static compilation work

It's still a gleam in the roadmap's eye, but I'm sure I could do a better job of handling -tags and your other flags than the go tool does, my feelings on the whole 'recompile your toolchain to disable cgo, etc' are well known.

> I'd also say that gb-vendor isn't going to work for us since it doesn't support using upstream release tags instead of hashes (even if you hand-munge the JSON), but since it's an optional piece I don't see that as a blocker myself.

That's a trivial fix to make, raise a bug and we will have it for you in a day.

**Comment by davecheney** [comment]
It's probably not going to help my case, but we now support `gb vendor fetch -tag tag URL`.

**Comment by tiborvass** [comment]
@davecheney thanks for helping out, really appreciated :)

Everyone,

Ultimately it boils down to adding a new maybe-popular-but-debated tool versus adding 30 lines of readable bash for something we need now.

I can't wait to ditch vendor.sh and I'm open to having a debate around vendoring solutions, but we need those lines of code out now, and I don't see consensus. This is a hot debate and I would also like to see what the Go team comes up with.

So I hope you'll still talk to me, but I'm bruteforcing by reopening the vendor.sh PR in order to get somewhere.

**Comment by tiborvass** [comment]
Oh and I thought I mentioned that Godep doesn't handle build tags, which is a big blocker:

```
$ cat main.go
// +build daemon

package main

import "github.com/Sirupsen/logrus"

func main() {
	logrus.WithFields(logrus.Fields{"Foo": "bar"}).Fatal()
}
$ godep save
godep: no buildable Go source files in /Users/tibor/go/src/github.com/tiborvass/helloworld
godep: error loading packages
```

Worse than that, if you mix files with build tags with files without build tags, only the ones without build tags are taken into account, making you think that it's all good and vendored when it's not.

**Comment by tiborvass** [comment]
FYI, Go 1.5 will have great progress on vendoring: https://groups.google.com/forum/#!msg/golang-dev/74zjMON9glU/4lWCRDCRZg0J and we can evaluate the right tool as all the vendoring tools migrate towards the new proposal.

**Comment by icecrime** [comment]
I'm glad we moved things forward. I'm sincerely saddened and frustrated that this was done without my participation, but purely by taking the initial idea from me.

Moving on.

**Comment by thaJeztah** [comment]
Noted, and I understand your frustration here, @icecrime. I'm happy to see things moving forward *initiated* by your ideas and work, but (even though without bad intent) I'd have rather seen this handled in a better way.

Drinks on me for you and @tianon at DockerCon

**Comment by aluzzardi** [comment]
> I can't wait to ditch vendor.sh and I'm open to having a debate around vendoring solutions, but we need those lines of code out now, and I don't see consensus. 

I don't really see why *we need those lines of code out now*. What was the urgency?

Our binaries are still the same size and our repo is still the same size - the only metric that effectively increased is our custom shell script complexity.

What's done is done - I guess we'll have a proper discussion once Go 1.5 hits the shelves.

**Pull request closed without merging**